opm/simulators/wells/WellFilterCake.cpp

```diff
@@ -37,106 +37,171 @@ namespace Opm {
+    applyCleaning(well, deferred_logger);
+}
 
-    const std::size_t np = well_state.numPhases();
-    for (int perf = 0; perf < well.numPerfs(); ++perf) {
-        // not considering the production water
-        const Scalar water_rates = std::max(Scalar{0.}, connection_rates[perf * np + water_index]);
-        const Scalar filtrate_rate = water_rates * conc;
-        filtration_particle_volume_[perf] += filtrate_rate * dt;
-        ws.perf_data.filtrate_data.rates[perf] = filtrate_rate;
-        ws.perf_data.filtrate_data.total[perf] = filtration_particle_volume_[perf];
+template<class Scalar>
+void WellFilterCake<Scalar>::
+applyCleaning(const WellInterfaceGeneric<Scalar>& well,
+              DeferredLogger& deferred_logger)
+{
+    const auto& connections = well.wellEcl().getConnections();
+    const auto nperf = well.numPerfs();
+    for (int perf = 0; perf < nperf; ++perf) {
+        const auto perf_ecl_index = well.perforationData()[perf].ecl_index;
+        const auto& connection = connections[perf_ecl_index];
+        if (!connection.filterCakeActive())
+            continue;
+
+        const auto& filter_cake = connection.getFilterCake();
+        const Scalar factor = filter_cake.sf_multiplier;
+        if (factor == 1.0)
+            continue;
+
+        filter_cake.sf_multiplier = 1.0;
+        skin_factor_[perf] *= factor;
+        const Scalar rw = connection.getFilterCakeRadius();
+        switch (filter_cake.geometry) {
+            case FilterCake::FilterCakeGeometry::LINEAR: {
+                // Previuos thickness adjusted to give correct cleaning multiplier at start of time step
+                thickness_[perf] *= factor;
+                break;
+            }
+            case FilterCake::FilterCakeGeometry::RADIAL: {
+                Scalar rc_prev = std::sqrt(rw * rw + 2. * rw * prev_thickness_[perf]);
+                // Previuos thickness and rc adjusted to give correct cleaning multiplier at start of time step
+                rc_prev = rw*std::exp(factor*std::log(rc_prev/rw));
+                thickness_[perf] = (rc_prev * rc_prev - rw * rw) / (2. * rw);
+                break;
+            }
+            default:
+                const auto geometry =
+                    FilterCake::filterCakeGeometryToString(filter_cake.geometry);
+                OPM_DEFLOG_THROW(std::runtime_error,
+                                    fmt::format(" Invalid filtration cake geometry type ({}) for well {}",
+                                                geometry, well.name()),
+                                    deferred_logger);
+        }
     }
 }
 
+
 template<class Scalar>
 void WellFilterCake<Scalar>::
-updateInjFCMult(const WellInterfaceGeneric<Scalar>& well,
-                WellState<Scalar>& well_state,
-                DeferredLogger& deferred_logger)
+updateMultipliers(const WellInterfaceGeneric<Scalar>& well,
```

Comment:
Yes, then preferably, the following code can be put in a function so we do not need to duplicate the same code in two places. 
```c++
        const auto denom = connection.ctfProperties().peaceman_denom;
        const auto denom2 = denom + skin_factor_[perf];
        inj_fc_multiplier_[perf] = denom / denom2;
```

After the code is put in a function, I do think it looks nicer that only call it once in the function `updatePreStep()`, in stead of calling it twice in both the function `applyCleaning()` (although it only happens occasionally) and `updateMultipliers()`. 

When `connection.ctfProperties().peaceman_denom` does not change, they are the same.  Then it is more about coding.

Reply:
But I do not insist because potentially `connection.ctfProperties().peaceman_denom` can change, then it becomes what we want physically.  

Wrapping the code in a function I think it is good for better readability and reducing code duplication.